### PR TITLE
[Merged by Bors] - feat: add list lemmas

### PR DESCRIPTION
**Pull request by sven-manthe**
Add some general purpose lemmas on lists. Also execute two adaptation notes.
The lemmas are taken from https://github.com/sven-manthe/A-formalization-of-Borel-determinacy-in-Lean

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Basic.lean [hunk 1]
```diff
@@ -1010,6 +1010,43 @@ theorem zipWith_flip (f : α → β → γ) : ∀ as bs, zipWith (flip f) bs as
...
+    · exact take_append_of_le_length h
+
+@[simp] lemma take_eq_right {x y : List α} n :
+  x.take n = (x ++ y).take n ↔ y = [] ∨ n ≤ x.length := by
+  rw [Eq.comm]; apply take_eq_left
+
+@[simp] lemma drop_take_append_drop (x : List α) (m n : ℕ) :
+  (x.drop m).take n ++ x.drop (n + m) = x.drop m := by rw [← drop_drop, take_append_drop]
+
+/--order of summands is swapped-/
```
Comment (grunweg) [inline Mathlib/Data/List/Basic.lean:1038]: I'd write this as
```suggestion
/-- Compared to `drop_take_append_drop`, the order of summands is swapped. -/
```
And I wonder: are really *both* lemmas useful, and should they both be simp lemmas?

File: Mathlib/Data/List/Basic.lean [hunk 2]
```diff
@@ -1010,6 +1010,43 @@ theorem zipWith_flip (f : α → β → γ) : ∀ as bs, zipWith (flip f) bs as
...
+@[simp] lemma drop_take_append_drop' (x : List α) (m n : ℕ) :
+  (x.drop m).take n ++ x.drop (m + n) = x.drop m := by rw [Nat.add_comm, drop_take_append_drop]
+
+/--take_concat_get in simp normal form-/
+@[simp] lemma take_concat_get' (l : List α) (i : ℕ) (h : i < l.length) :
```
Comment (grunweg) [inline Mathlib/Data/List/Basic.lean:1043]: This makes me wonder if `take_concat_get` should be changed instead.

File: Mathlib/Data/List/Basic.lean [hunk 3]
```diff
@@ -1010,6 +1010,43 @@ theorem zipWith_flip (f : α → β → γ) : ∀ as bs, zipWith (flip f) bs as
...
 
+@[simp] lemma take_eq_self (x : List α) n : x.take n = x ↔ x.length ≤ n :=
+  ⟨fun h ↦ by rw [← h]; simp; omega, take_of_length_le⟩
+
+@[simp] lemma take_self_eq (x : List α) n : x = x.take n ↔ x.length ≤ n := by
+  rw [Eq.comm, take_eq_self]
+
+@[simp] lemma take_eq_left {x y : List α} n :
+  (x ++ y).take n = x.take n ↔ y = [] ∨ n ≤ x.length := by
+  constructor
```
Comment (grunweg) [inline Mathlib/Data/List/Basic.lean:1021]: I have the glaring feeling this proof can be optimised. I'm not familiar enough with the library around lists to quickly suggest how, though. Perhaps another reviewer can provide some feedback.

File: Mathlib/Algebra/BigOperators/Group/List.lean [hunk 4]
```diff
@@ -205,7 +205,7 @@ theorem prod_take_mul_prod_drop (L : List M) (i : ℕ) :
 @[to_additive (attr := simp)]
 theorem prod_take_succ (L : List M) (i : ℕ) (p : i < L.length) :
     (L.take (i + 1)).prod = (L.take i).prod * L[i] := by
-  simp [take_succ, p]
+  simp [← take_concat_get', p]
```
Comment (grunweg) [inline Mathlib/Algebra/BigOperators/Group/List.lean:208]: Do you have an idea why this change is necessary?

**Review by grunweg — COMMENTED** [review]
Thanks! I have a few style comments and a few questions. I don't know this part of mathlib well - so take my comments around naming and library design with a pinch of salt.

**Comment by grunweg** [comment]
(If you want, you could certainly split this into several PRs, for instance one on the adaptation notes and another for the `List/Infix` changes. I would approve both right away. This might speed up reviews, but I am aware this may cause additional overhead for you.)

File: Mathlib/Data/List/Infix.lean [hunk 5]
```diff
@@ -51,6 +51,19 @@ theorem eq_of_prefix_of_length_eq (h : l₁ <+: l₂) : l₁.length = l₂.lengt
...
+  simpa [prefix_take_iff, Nat.min_le_left] using (take_prefix n l₁).trans h
+
+lemma IsPrefix.drop (h : l₁ <+: l₂) n : l₁.drop n <+: l₂.drop n := by
+  rw [prefix_iff_eq_take.mp h, drop_take]; apply take_prefix
+
+lemma isPrefix_append_of_length (h : l₁.length ≤ l₂.length) : l₁ <+: l₂ ++ l₃ ↔ l₁ <+: l₂ :=
+  ⟨fun h ↦ by rw [prefix_iff_eq_take] at *; nth_rw 1 [h, take_eq_left]; tauto,
+   fun h ↦ h.trans <| l₂.prefix_append l₃⟩
+
+@[simp] lemma take_prefix_take {m n} : l.take m <+: l.take n ↔ m ≤ n ∨ l.length ≤ n := by
```
[the remark this answers is not in the document]
Comment (b-mehta) [reply Mathlib/Data/List/Infix.lean:64]: `take_isPrefix_take_iff`, maybe?

[the remark this answers is not in the document]
Comment (sven-manthe) [reply Mathlib/Data/List/Infix.lean:64]: Since `take_eq_take` already exists, I'll use `take_isPrefix_take`

File: Mathlib/Data/List/Basic.lean (hunk 2 above)
Comment (sven-manthe) [reply Mathlib/Data/List/Basic.lean:1043]: I would have done this if it were part of mathlib and not core

File: Mathlib/Algebra/BigOperators/Group/List.lean (hunk 4 above)
Comment (sven-manthe) [reply Mathlib/Algebra/BigOperators/Group/List.lean:208]: `take_succ` and `Iff.symm take_concat_get'` are almost the same lemma. Thus removing the new global simp lemma `take_concat_get'` from the local simp set and leaving `take_succ` in there also works, but the old simp set looped.

File: Mathlib/Data/List/Basic.lean (hunk 3 above)
Comment (sven-manthe) [reply Mathlib/Data/List/Basic.lean]: Your feeling was right

File: Mathlib/Data/List/Basic.lean [hunk 6]
```diff
@@ -1010,6 +1010,35 @@ theorem zipWith_flip (f : α → β → γ) : ∀ as bs, zipWith (flip f) bs as
...
+  rw [Eq.comm]; apply take_eq_left_iff
+
+@[simp] lemma drop_take_append_drop (x : List α) (m n : ℕ) :
+    (x.drop m).take n ++ x.drop (n + m) = x.drop m := by rw [← drop_drop, take_append_drop]
+
+/-- Compared to `drop_take_append_drop`, the order of summands is swapped. -/
+@[simp] lemma drop_take_append_drop' (x : List α) (m n : ℕ) :
+    (x.drop m).take n ++ x.drop (m + n) = x.drop m := by rw [Nat.add_comm, drop_take_append_drop]
+
+/--take_concat_get in simp normal form-/
```
Comment (IvanRenison) [inline Mathlib/Data/List/Basic.lean:1034]:
```suggestion
/-- `take_concat_get` in simp normal form -/
```

File: Mathlib/Data/List/Basic.lean [hunk 7]
```diff
@@ -1010,6 +1010,35 @@ theorem zipWith_flip (f : α → β → γ) : ∀ as bs, zipWith (flip f) bs as
...
+@[simp] lemma take_concat_get' (l : List α) (i : ℕ) (h : i < l.length) :
+  l.take i ++ [l[i]] = l.take (i + 1) := by simpa using take_concat_get l i h
+
+/--eq_nil_or_concat in simp normal form-/
```
Comment (IvanRenison) [inline Mathlib/Data/List/Basic.lean:1038]:
```suggestion
/-- `eq_nil_or_concat` in simp normal form -/
```

**Review by fpvandoorn — COMMENTED** [review]
bors d+
